### PR TITLE
Limit the number of commits in completions

### PR DESCRIPTION
**Pull request by gladhorn**
Offering auto completion for existing commits is great, but on big
repositories, it suddenly becomes really slow, even with fast hard
disks, since each commit is read and then a line processed for it.

Instead limit to the last 500 commits (arbitrary number) which still
feels fast. Going back further in history can easily and more reasonably
done with git log etc.

### CODE REVIEW

**Comment by krader1961** [comment]
👍 Seems reasonable to me. Let's let this bake for a couple of days to see if anyone objects.

**Comment by floam** [comment]
How do the official bash completions do it (or zsh's best ones)? Have they found a more clever way to narrow the scope of the results returned by `git` (perhaps in line with removing revnos that aren't particularly actionable given the context?)

**Comment by gladhorn** [comment]
I don't think that the official bash completion completes commits - offering the sha1s with the short message is a big luxury that fish offers ;)

**Comment by gladhorn** [comment]
Of course the 500 is a crazy random number - but it makes is fast enough for my purposes, I imagine it's still slow for people without SSD. And going back that far is of course crazy anyway, then you're better off doing it by other means (git log with search etc).

**Comment by faho** [comment]
Here's the problem I have with this approach: If you enter a part of a commit hash and it's not among those randomly (is it really or does git pick the last 500?) selected, you won't get a completion.

For `cherry-pick`, I added the `__fish_git_possible_commithash` [function](https://github.com/gladhorn/fish-shell/blob/6078473dc5ab63c9011d697ff15c241a953a7fa4/share/completions/git.fish#L194) that returns true when you entered at least 3 characters that could be a hash (i.e. are valid hex-chars), and that strikes me to be a better solution - when you entered a sha1, you'll get your sha1 completed.

Though I haven't tested this on huge repos - which one are you testing on, @gladhorn?

**Comment by gladhorn** [comment]
My day to day repos are git://code.qt.io/qt/qtbase.git (29396 commits) and other Qt repositories.
What I really like is the current version in context of fixup. Assuming you work on a series of patches and want to change a detail in the first patch. Make sure to have autosquash in your git config. Now create a new commit for the change of the original patch with "git commit --fixup=" and use autocompletion to find the first commit that you want to change (you have the commit short title to help you), press return, run rebase and done.

So in short, listing the most recent commits for me is invaluable.
I see how this conflicts/is alternative to the completion of sha1s that have been typed manually.

**Comment by gladhorn** [comment]
About the random part: git picks the most recent 500 commits. I'm not sure how it decides among branches. For most people it should return the right commits though :)

**Comment by faho** [comment]
>My day to day repos are git://code.qt.io/qt/qtbase.git (29396 commits) and other Qt repositories.

Neat, that's something I use daily.

I cloned qtbase (which has about 5 times as many commits as fish) and completing `git show` takes about 3 seconds for me. With "--max-count=5000" (i.e. ten times more than in this PR), it takes about half a second. "500" is imperceptible. My machine is probably on the low end.

@gladhorn: One possibility I've thought about is limiting only for select commands, like the fixup case you mentioned. Because you're not gonna fixup a commit that's a year old, but you might want to `show` it. Would that make sense to you?

**Comment by gladhorn** [comment]
Looking at various git commands, it seems not very straight forward to construct something that looks for sha1 and short message at the same time. Therefore I think the current approach is probably quite good for the use case of searching the entire history. If it was only to find messages, git log --grep would be good, for only finding sha1s maybe git rev-parse --disambiguate=sha1 (needs four hex-chars), even though I had trouble getting sensible results with that.

So I completely agree, splitting it into two functions with different goal, or having a parameter to limit the number of commits makes sense. No need to over do it either :)
I can have a look at how the function is used soonish and update this patch.

**Comment by krader1961** [comment]
@gladhorn, This is starting to smell of neglect. Personally I'm happy with the behavior this change introduces. But then I'm a neophyte git user who might not notice problems this introduces. In your opinion should this be merged? If not, even if you intend to further refine the concept, we should close this particular PR.

**Comment by gladhorn** [comment]
Is there any way to sort completions? For fixup, I'd like to return commits in the current branch, and then it's most sensible to return the most recent ones first.

**Comment by gladhorn** [comment]
Or rather the opposite - to return a list of completions that is then not sorted later on. Because git returns them in the right order.

**Comment by faho** [comment]
>Is there any way to sort completions? 

@gladhorn: No, see #361.

**Comment by krader1961** [comment]
@gladhorn, What's the status of this PR? It's been three weeks since the last update. Should it be merged? If you're still working on this change and have no idea when it will be ready for another review this PR should be closed. You can open a new PR when you're ready for us to take another look.

**Comment by gladhorn** [comment]
@krader1961 to me this is still a clear improvement. As long as there is no sorting of completions beyond the default sorting it's probably as good as it gets.

**Comment by faho** [comment]
@gladhorn: Merged. It'd be nice to figure out how to handle this properly in the general case, but that's gonna take some more work.

**Comment by lgarron** [comment]
I commented on https://github.com/fish-shell/fish-shell/commit/f37995c676c7974b18dd0bca0365506d335f18fc#commitcomment-18822542 but it seems the commit is `50`, while this issue and the commit description say `500`. Not sure which was intended.

(Doesn't make a difference for my problems in #3342 either way.)
